packages/gatsby-link/src/index.tsx

```diff
@@ -182,39 +261,55 @@ class GatsbyLink extends React.Component {
   }
 }
 
-GatsbyLink.propTypes = {
-  ...NavLinkPropTypes,
-  onClick: PropTypes.func,
-  to: PropTypes.string.isRequired,
-  replace: PropTypes.bool,
-  state: PropTypes.object,
-}
-
-const showDeprecationWarning = (functionName, altFunctionName, version) =>
+const showDeprecationWarning = (
+  functionName: string,
+  altFunctionName: string,
+  version: number
+): void =>
   console.warn(
     `The "${functionName}" method is now deprecated and will be removed in Gatsby v${version}. Please use "${altFunctionName}" instead.`
   )
 
-export default React.forwardRef((props, ref) => (
-  <GatsbyLink innerRef={ref} {...props} />
+/**
+ * This component is intended _only_ for links to pages handled by Gatsby. For links to pages on other
+ * domains or pages on the same domain not handled by the current Gatsby site, use the normal `<a>` element.
+ */
+const GatsbyLink = React.forwardRef<
+  HTMLAnchorElement,
+  GatsbyLinkProps<unknown>
```

Comment:
don't we need to pass in a generic here for it to ever resolve to anything in the actual `GatsbyLinkInternal` functions?

Reply:
I think since we are explicitly typing the arguments there isn't a benefit to passing in the generics

Reply:
In TS var/const/let variables cannot be declared as generic, so there's no way to parameterize the `const GatsbyLink`. The `type GatsbyLink` below achieves parameterizing this through [TypeScript declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#merging-interfaces).

The mean reason I'm doing this is because the old index.d.ts declared GatsbyLink as

```
export class GastbyLink<TState> extends React.Component<....>
```